File: internal/sources/shared/git.go
package shared

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

// GetHeadSha returns the current HEAD SHA for the given project path.
func GetHeadSha(projectPath string) string {
	if projectPath == "" {
		return ""
	}
	out, err := exec.Command("git", "-C", projectPath, "rev-parse", "HEAD").Output()
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(out))
}

// GetBranch returns the current git branch name for the given project path.
func GetBranch(projectPath string) string {
	if projectPath == "" {
		return ""
	}
	out, err := exec.Command("git", "-C", projectPath, "rev-parse", "--abbrev-ref", "HEAD").Output()
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(out))
}

// GetGitDiff returns the combined diff of tracked changes and new untracked
// files for the given project path. Output is truncated to 50KB.
func GetGitDiff(projectPath string) string {
	if projectPath == "" {
		return ""
	}
	cmd := exec.Command("git", "diff", "HEAD")
	cmd.Dir = projectPath
	tracked, _ := cmd.Output()

	untracked := untrackedDiff(projectPath)

	combined := string(tracked) + untracked
	if combined == "" {
		return ""
	}
	const maxBytes = 50_000
	if len(combined) > maxBytes {
		return combined[:maxBytes] + "\n[truncated]"
	}
	return combined
}

// untrackedDiff generates a unified-diff representation of untracked new files.
func untrackedDiff(projectPath string) string {
	out, err := exec.Command("git", "-C", projectPath, "status", "--porcelain").Output()
	if err != nil || len(out) == 0 {
		return ""
	}
	var sb strings.Builder
	for _, line := range strings.Split(string(out), "\n") {
		if len(line) < 4 || line[:2] != "??" {
			continue
		}
		rel := strings.TrimSpace(line[3:])
		if rel == "" || strings.HasSuffix(rel, "/") {
			continue
		}
		content, err := os.ReadFile(filepath.Join(projectPath, rel))
		if err != nil {
			continue
		}
		check := content
		if len(check) > 8192 {
			check = check[:8192]
		}
		if strings.ContainsRune(string(check), 0) {
			continue
		}
		lines := strings.Split(strings.TrimRight(string(content), "\n"), "\n")
		fmt.Fprintf(&sb, "diff --git a/%s b/%s\nnew file mode 100644\n--- /dev/null\n+++ b/%s\n@@ -0,0 +1,%d @@\n",
			rel, rel, rel, len(lines))
		for _, l := range lines {
			fmt.Fprintf(&sb, "+%s\n", l)
		}
	}
	return sb.String()
}

// GetCommitsSince returns commit SHAs after the given ISO timestamp.
func GetCommitsSince(projectPath, sinceISO string) []string {
	cmd := exec.Command("git", "log", "--format=%H", "--after="+sinceISO)
	cmd.Dir = projectPath
	out, err := cmd.Output()
	if err != nil {
		return nil
	}
	return FilterNonEmpty(strings.Split(strings.TrimSpace(string(out)), "\n"))
}

// GetCommitRange returns commit SHAs between two unix-millisecond timestamps.
func GetCommitRange(projectPath string, sinceMs, untilMs *int64) []string {
	args := []string{"log", "--format=%H", "--no-merges"}
	if sinceMs != nil {
		t := fmt.Sprintf("%d", *sinceMs/1000)
		args = append(args, "--after=@"+t)
	}
	if untilMs != nil {
		t := fmt.Sprintf("%d", *untilMs/1000)
		args = append(args, "--before=@"+t)
	}
	cmd := exec.Command("git", args...)
	cmd.Dir = projectPath
	out, err := cmd.Output()
	if err != nil {
		return nil
	}
	return FilterNonEmpty(strings.Split(strings.TrimSpace(string(out)), "\n"))
}

// FilterNonEmpty removes empty/whitespace-only strings from a slice.
func FilterNonEmpty(lines []string) []string {
	var result []string
	for _, l := range lines {
		if strings.TrimSpace(l) != "" {
			result = append(result, l)
		}
	}
	return result
}

// TouchedProjectIDs returns all registered project IDs whose files appear in
// tool call paths.
func TouchedProjectIDs(toolCalls []map[string]any, projByID map[string]string) []string {
	seen := map[string]bool{}
	for _, tc := range toolCalls {
		path := extractPathFromToolCall(tc)
		if path == "" {
			continue
		}
		for id, projPath := range projByID {
			if projPath != "" && strings.HasPrefix(path, projPath+"/") {
				seen[id] = true
			}
		}
	}
	ids := make([]string, 0, len(seen))
	for id := range seen {
		ids = append(ids, id)
	}
	return ids
}

// RepoSnapshots returns git snapshots for repos OTHER than primaryProjectID
// that are referenced by tool call file paths.
func RepoSnapshots(toolCalls []map[string]any, primaryProjectID string, projByID map[string]string) map[string]any {
	result := map[string]any{}
	for _, tc := range toolCalls {
		path := extractPathFromToolCall(tc)
		if path == "" {
			continue
		}
		for id, projPath := range projByID {
			if id == primaryProjectID || projPath == "" {
				continue
			}
			if strings.HasPrefix(path, projPath+"/") {
				if _, ok := result[id]; !ok {
					result[id] = map[string]any{
						"head_sha": GetHeadSha(projPath),
						"git_diff": GetGitDiff(projPath),
					}
				}
			}
		}
	}
	if len(result) == 0 {
		return nil
	}
	return result
}

// ChangedFilesFromToolCalls extracts file paths from write-oriented tool calls.
func ChangedFilesFromToolCalls(toolCalls []map[string]any) []string {
	writeTools := map[string]bool{
		"write_file":              true,
		"create_file":             true,
		"edit_file":               true,
		"replace_string_in_file":  true,
		"multi_replace_string_in_file": true,
		"edit_notebook_file":      true,
		"Write":                   true, // Claude Code tool name
	}

	seen := map[string]bool{}
	var files []string
	for _, tc := range toolCalls {
		tool, _ := tc["tool"].(string)
		if !writeTools[tool] {
			continue
		}
		path := extractPathFromToolCall(tc)
		if path != "" && !seen[path] {
			seen[path] = true
			files = append(files, path)
		}
	}
	return files
}

// extractPathFromToolCall extracts a file path from a tool call's input.
func extractPathFromToolCall(tc map[string]any) string {
	if input, ok := tc["input"].(map[string]any); ok {
		for _, key := range []string{"path", "file_path", "filePath"} {
			if p, ok := input[key].(string); ok && p != "" {
				return p
			}
		}
	}
	// Fallback: top-level path
	if p, ok := tc["path"].(string); ok && p != "" {
		return p
	}
	return ""
}

File: crates/pcr-core/src/sources/shared/git.rs
//! Git wrappers. Direct port of `cli/internal/sources/shared/git.go`.
//!
//! Each function shells out to the `git` binary exactly like the Go version.
//! On Windows, `git.exe` lives in `C:\Program Files\Git\` which is an
//! AppLocker-allowed path, so spawning it from inside `node.exe` (for the
//! napi build) or from a standalone binary (for the brew build) always
//! passes AppLocker path rules.

use std::path::Path;
use std::process::Command;

pub fn get_head_sha(project_path: &str) -> String {
    if project_path.is_empty() {
        return String::new();
    }
    run(&["-C", project_path, "rev-parse", "HEAD"], None)
}

/// Returns the current branch name for `project_path`, or empty string
/// when the working tree is in a detached-HEAD state.
///
/// `git rev-parse --abbrev-ref HEAD` returns the literal string "HEAD"
/// when detached, which downstream display / push code would mistake for
/// a real branch named "HEAD". Filter it here once instead of at every
/// call site.
pub fn get_branch(project_path: &str) -> String {
    if project_path.is_empty() {
        return String::new();
    }
    let raw = run(
        &["-C", project_path, "rev-parse", "--abbrev-ref", "HEAD"],
        None,
    );
    if raw == "HEAD" {
        String::new()
    } else {
        raw
    }
}

/// Returns true when `project_path` is inside a git working tree. Used by
/// the watchers to distinguish "no diff because there's nothing to diff"
/// from "no diff because git isn't available". The latter ends up tagged
/// in `file_context.git_unavailable: true` so reviewers see why the diff
/// is empty instead of being misled.
///
/// Empty paths return false (vacuous — not even attempted). A successful
/// `git rev-parse --is-inside-work-tree` writes "true" to stdout.
///
/// Result is cached per-process — projects don't move between
/// "is a git repo" and "isn't a git repo" within a watcher's lifetime,
/// and the watchers call this on every save.
pub fn is_git_repo(project_path: &str) -> bool {
    if project_path.is_empty() {
        return false;
    }
    if let Some(cached) = lookup_is_git_repo_cache(project_path) {
        return cached;
    }
    let out = run(
        &["-C", project_path, "rev-parse", "--is-inside-work-tree"],
        None,
    );
    let result = out.trim() == "true";
    insert_is_git_repo_cache(project_path.to_string(), result);
    result
}

fn is_git_repo_cache() -> &'static std::sync::Mutex<std::collections::HashMap<String, bool>> {
    static CACHE: std::sync::OnceLock<std::sync::Mutex<std::collections::HashMap<String, bool>>> =
        std::sync::OnceLock::new();
    CACHE.get_or_init(|| std::sync::Mutex::new(std::collections::HashMap::new()))
}

fn lookup_is_git_repo_cache(key: &str) -> Option<bool> {
    is_git_repo_cache().lock().ok()?.get(key).copied()
}

fn insert_is_git_repo_cache(key: String, value: bool) {
    if let Ok(mut g) = is_git_repo_cache().lock() {
        g.insert(key, value);
    }
}

/// `git diff HEAD` combined with a synthetic diff of untracked files.
/// Capped at 50 KB. Matches `shared.GetGitDiff`.
pub fn get_git_diff(project_path: &str) -> String {
    if project_path.is_empty() {
        return String::new();
    }
    let tracked = run(&["diff", "HEAD"], Some(project_path));
    let untracked = untracked_diff(project_path);
    let combined = format!("{tracked}{untracked}");
    if combined.is_empty() {
        return String::new();
    }
    const MAX: usize = 50_000;
    if combined.len() > MAX {
        let mut out = combined[..MAX].to_string();
        out.push_str("\n[truncated]");
        out
    } else {
        combined
    }
}

fn untracked_diff(project_path: &str) -> String {
    // NUL-terminated porcelain so filenames with spaces, quotes, or
    // newlines round-trip verbatim. Without `-z`, git shell-escapes
    // problem paths and a naive `line[3..].trim()` parser would
    // produce literal `"foo bar.rs"` strings that fail every later
    // file read.
    let cmd = std::process::Command::new("git")
        .args(["-C", project_path, "status", "--porcelain=v1", "-z"])
        .output();
    let Ok(output) = cmd else {
        return String::new();
    };
    if !output.status.success() || output.stdout.is_empty() {
        return String::new();
    }
    let mut sb = String::new();
    let bytes = &output.stdout;
    for field in bytes.split(|b| *b == 0) {
        if field.len() < 4 || &field[..2] != b"??" {
            continue;
        }
        let Ok(rel) = std::str::from_utf8(&field[3..]) else {
            continue;
        };
        if rel.is_empty() || rel.ends_with('/') {
            continue;
        }
        let Ok(content) = std::fs::read(Path::new(project_path).join(rel)) else {
            continue;
        };
        let check_len = content.len().min(8192);
        if content[..check_len].contains(&0) {
            continue;
        }
        let s = String::from_utf8_lossy(&content);
        let trimmed = s.trim_end_matches('\n');
        let lines: Vec<&str> = trimmed.split('\n').collect();
        sb.push_str(&format!(
            "diff --git a/{rel} b/{rel}\nnew file mode 100644\n--- /dev/null\n+++ b/{rel}\n@@ -0,0 +1,{} @@\n",
            lines.len()
        ));
        for l in lines {
            sb.push_str(&format!("+{l}\n"));
        }
    }
    sb
}

/// `git log --format=%H --after=<sinceISO>` from inside `project_path`.
pub fn get_commits_since(project_path: &str, since_iso: &str) -> Vec<String> {
    let out = run(
        &["log", "--format=%H", &format!("--after={since_iso}")],
        Some(project_path),
    );
    filter_non_empty(out.trim().split('\n'))
}

/// `git log --format=%H --no-merges` with optional millisecond epoch bounds.
pub fn get_commit_range(
    project_path: &str,
    since_ms: Option<i64>,
    until_ms: Option<i64>,
) -> Vec<String> {
    let mut args: Vec<String> = vec!["log".into(), "--format=%H".into(), "--no-merges".into()];
    if let Some(ms) = since_ms {
        args.push(format!("--after=@{}", ms / 1000));
    }
    if let Some(ms) = until_ms {
        args.push(format!("--before=@{}", ms / 1000));
    }
    let out = run_owned(&args, Some(project_path));
    filter_non_empty(out.trim().split('\n'))
}

pub fn filter_non_empty<'a, I: IntoIterator<Item = &'a str>>(it: I) -> Vec<String> {
    it.into_iter()
        .filter(|l| !l.trim().is_empty())
        .map(|s| s.to_string())
        .collect()
}

fn run(args: &[&str], cwd: Option<&str>) -> String {
    let mut cmd = Command::new("git");
    cmd.args(args);
    if let Some(cwd) = cwd {
        cmd.current_dir(cwd);
    }
    match cmd.output() {
        Ok(o) if o.status.success() => String::from_utf8_lossy(&o.stdout).trim_end().to_string(),
        _ => String::new(),
    }
}

fn run_owned(args: &[String], cwd: Option<&str>) -> String {
    let mut cmd = Command::new("git");
    cmd.args(args);
    if let Some(cwd) = cwd {
        cmd.current_dir(cwd);
    }
    match cmd.output() {
        Ok(o) if o.status.success() => String::from_utf8_lossy(&o.stdout).trim_end().to_string(),
        _ => String::new(),
    }
}

/// Wrapper to run a git subcommand and return trimmed stdout. Public for
/// commands that need their own git calls (push, init, log, etc.).
pub fn git_output(args: &[&str]) -> String {
    run(args, None)
}

pub fn git_output_in(dir: &str, args: &[&str]) -> String {
    run(args, Some(dir))
}
